driver/src/main/java/org/neo4j/driver/internal/cursor/ResultCursorImpl.java

```diff
@@ -0,0 +1,1383 @@
+import org.neo4j.driver.exceptions.Neo4jException;
+import org.neo4j.driver.exceptions.NoSuchRecordException;
+import org.neo4j.driver.internal.DatabaseBookmark;
+import org.neo4j.driver.internal.FailableCursor;
+import org.neo4j.driver.internal.InternalRecord;
+import org.neo4j.driver.internal.async.UnmanagedTransaction;
+import org.neo4j.driver.internal.bolt.api.BoltConnection;
+import org.neo4j.driver.internal.bolt.api.BoltProtocolVersion;
+import org.neo4j.driver.internal.bolt.api.GqlStatusError;
+import org.neo4j.driver.internal.bolt.api.ResponseHandler;
+import org.neo4j.driver.internal.bolt.api.summary.BeginSummary;
+import org.neo4j.driver.internal.bolt.api.summary.DiscardSummary;
+import org.neo4j.driver.internal.bolt.api.summary.PullSummary;
+import org.neo4j.driver.internal.bolt.api.summary.RunSummary;
+import org.neo4j.driver.internal.bolt.api.summary.TelemetrySummary;
+import org.neo4j.driver.internal.telemetry.ApiTelemetryWork;
+import org.neo4j.driver.internal.util.Futures;
+import org.neo4j.driver.internal.util.MetadataExtractor;
+import org.neo4j.driver.summary.ResultSummary;
+
+public class ResultCursorImpl extends AbstractRecordStateResponseHandler
+        implements ResultCursor, FailableCursor, ResponseHandler {
+    public static final MetadataExtractor METADATA_EXTRACTOR = new MetadataExtractor("t_last");
+    private static final ClientException IGNORED_ERROR = new ClientException(
+            GqlStatusError.UNKNOWN.getStatus(),
+            GqlStatusError.UNKNOWN.getStatusDescription("A message has been ignored during result streaming."),
+            "N/A",
+            "A message has been ignored during result streaming.",
+            GqlStatusError.DIAGNOSTIC_RECORD,
+            null);
+    private final BoltConnection boltConnection;
+    private final Queue<Record> records = new ArrayDeque<>();
+    private final Query query;
+    private final long fetchSize;
+    private final Consumer<Throwable> throwableConsumer;
+    private final Consumer<DatabaseBookmark> bookmarkConsumer;
+    private final Supplier<Throwable> termSupplier;
+    private final boolean closeOnSummary;
+    private final boolean legacyNotifications;
+    private final CompletableFuture<ResultCursorImpl> resultCursorFuture = new CompletableFuture<>();
+    private final CompletableFuture<UnmanagedTransaction> beginFuture;
+    private final ApiTelemetryWork apiTelemetryWork;
+    private final CompletableFuture<Void> consumedFuture = new CompletableFuture<>();
+    private RunSummary runSummary;
+    private State state;
+
+    private boolean apiCallInProgress;
+    private CompletableFuture<Record> peekFuture;
+    private CompletableFuture<Record> recordFuture;
+    private CompletableFuture<Boolean> secondRecordFuture;
+    private CompletableFuture<List<Record>> recordsFuture;
+    private boolean keepRecords;
+    private CompletableFuture<ResultSummary> summaryFuture;
+    private ResultSummary summary;
```

Comment:
can be final

Reply:
With this implementation, these are mutable.